File: app2/src/main/java/ru/neverlands/anclient/manager/ProfessionRatingMonitor.java
package ru.neverlands.anclient.manager;

import android.content.Context;
import android.content.SharedPreferences;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import ru.neverlands.anclient.info.ProfessionRatingRepository;
import ru.neverlands.anclient.postfilter.MainPhp;
import ru.neverlands.anclient.utils.AppLog;
import ru.neverlands.anclient.utils.AppVars;
import ru.neverlands.anclient.utils.Chat;

/**
 * Проверяет воскресные Top 10 weekly-рейтингов и публикует локальные чат-уведомления.
 */
public final class ProfessionRatingMonitor {
    private static final String TAG = "ProfessionRatingMonitor";
    private static final String TRACE_CHAIN = "PROF_RATING_TRACE";
    private static final String PREFS_NAME = "profession_rating_monitor";
    private static final String KEY_LAST_CHECK_HOUR = "last_check_hour";
    private static final long RETRY_AFTER_FAILURE_MS = 10 * 60 * 1000L;

    private static volatile boolean checkRunning;
    private static volatile long lastCheckStartedAtMs;

    private ProfessionRatingMonitor() {
    }

    public static void maybeCheck(Context context) {
        if (context == null || AppVars.Profile == null) {
            return;
        }
        String nick = resolveCurrentNick();
        if (nick.isEmpty()) {
            return;
        }

        Date serverDate = resolveServerDate();
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(serverDate);
        if (calendar.get(Calendar.DAY_OF_WEEK) != Calendar.SUNDAY || calendar.get(Calendar.HOUR_OF_DAY) < 15) {
            return;
        }

        String hourKey = new SimpleDateFormat("yyyyMMddHH", Locale.US).format(serverDate);
        String preferenceKey = KEY_LAST_CHECK_HOUR + "_"
                + Integer.toHexString(ProfessionRatingRepository.normalizeNick(nick).hashCode());
        SharedPreferences prefs = context.getApplicationContext()
                .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        if (hourKey.equals(prefs.getString(preferenceKey, ""))) {
            return;
        }

        long now = System.currentTimeMillis();
        if (checkRunning || now - lastCheckStartedAtMs < RETRY_AFTER_FAILURE_MS) {
            return;
        }

        checkRunning = true;
        lastCheckStartedAtMs = now;
        Context appContext = context.getApplicationContext();
        new Thread(() -> runCheck(appContext, nick, hourKey, preferenceKey), "prof-rating-monitor").start();
    }

    private static void runCheck(Context context, String currentNick, String hourKey, String preferenceKey) {
        boolean anySuccess = false;
        int hitCount = 0;
        try {
            String normalizedNick = ProfessionRatingRepository.normalizeNick(currentNick);
            List<ProfessionRatingRepository.Category> categories = ProfessionRatingRepository.getCategories();
            for (ProfessionRatingRepository.Category category : categories) {
                try {
                    ProfessionRatingRepository.RatingTable table = ProfessionRatingRepository.loadRating(category.id, true);
                    anySuccess = true;
                    ProfessionRatingRepository.RatingEntry hit = findTopTenHit(table, normalizedNick);
                    if (hit != null) {
                        hitCount++;
                        postHitNotification(category, hit);
                    }
                } catch (Exception error) {
                    AppLog.w(TRACE_CHAIN, TAG, "WEEKLY_RATING_CHECK_FAILED: id=" + category.id
                            + ", title=" + category.title, error);
                }
            }
            if (anySuccess) {
                context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                        .edit()
                        .putString(preferenceKey, hourKey)
                        .apply();
            }
            AppLog.i(TRACE_CHAIN, TAG, "WEEKLY_RATING_CHECK_DONE: nick=" + currentNick
                    + ", hour=" + hourKey + ", hits=" + hitCount + ", anySuccess=" + anySuccess);
        } finally {
            checkRunning = false;
        }
    }

    private static ProfessionRatingRepository.RatingEntry findTopTenHit(
            ProfessionRatingRepository.RatingTable table,
            String normalizedNick) {
        if (table == null || normalizedNick.isEmpty()) {
            return null;
        }
        int limit = Math.min(10, table.entries.size());
        for (int i = 0; i < limit; i++) {
            ProfessionRatingRepository.RatingEntry entry = table.entries.get(i);
            if (normalizedNick.equals(ProfessionRatingRepository.normalizeNick(entry.nick))) {
                return entry;
            }
        }
        return null;
    }

    private static void postHitNotification(
            ProfessionRatingRepository.Category category,
            ProfessionRatingRepository.RatingEntry entry) {
        String message = MainPhp.buildServerChatTimeHtmlExternal()
                + "<font color=#333399><b>[ANClient]</b>:</font> "
                + "Вы попали в рейтинг " + escapeHtml(category.title)
                + " под № " + entry.rank
                + ". Не забудьте забрать награду!";
        Chat.addMessageToChat(message);
        AppLog.i(TRACE_CHAIN, TAG, "WEEKLY_RATING_TOP10_NOTIFY: category=" + category.title
                + ", rank=" + entry.rank + ", nick=" + entry.nick);
    }

    private static String resolveCurrentNick() {
        String nick = AppVars.Profile != null && AppVars.Profile.UserNick != null
                ? AppVars.Profile.UserNick.trim()
                : "";
        if (nick.isEmpty()) {
            nick = AppVars.RuntimeAuthUserNick == null ? "" : AppVars.RuntimeAuthUserNick.trim();
        }
        return nick;
    }

    private static Date resolveServerDate() {
        long serverMs = System.currentTimeMillis();
        if (AppVars.Profile != null && AppVars.Profile.ServDiff != Long.MIN_VALUE) {
            serverMs -= AppVars.Profile.ServDiff;
        } else if (AppVars.ServerDateTime != null) {
            serverMs = AppVars.ServerDateTime.getTime();
        }
        return new Date(serverMs);
    }

    private static String escapeHtml(String value) {
        if (value == null) {
            return "";
        }
        return value.replace("&", "&amp;")
                .replace("<", "&lt;")
                .replace(">", "&gt;")
                .replace("\"", "&quot;");
    }
}

File: ANClient/Info/ProfessionRatingMonitor.cs
namespace ANClient.Info
{
    using System;
    using System.Collections.Generic;
    using System.Globalization;
    using System.Threading;

    internal static class ProfessionRatingMonitor
    {
        private const string Tag = "ProfessionRatingMonitor";
        private const string TraceChain = "prof_rating_trace";
        private static readonly object SyncRoot = new object();
        private static readonly Dictionary<string, string> LastCheckedHourByNick = new Dictionary<string, string>();
        private static readonly TimeSpan RetryAfterFailure = TimeSpan.FromMinutes(10);

        private static bool _checkRunning;
        private static DateTime _lastCheckStartedAtUtc = DateTime.MinValue;

        internal static void MaybeCheck()
        {
            if (ANClient.AppVars.Profile == null)
            {
                return;
            }

            string nick = ResolveCurrentNick();
            if (nick.Length == 0)
            {
                return;
            }

            DateTime serverDate = ResolveServerDate();
            if (serverDate.DayOfWeek != DayOfWeek.Sunday || serverDate.Hour < 15)
            {
                return;
            }

            string hourKey = serverDate.ToString("yyyyMMddHH", CultureInfo.InvariantCulture);
            string nickKey = ProfessionRatingRepository.NormalizeNick(nick);
            lock (SyncRoot)
            {
                string lastHour;
                if (LastCheckedHourByNick.TryGetValue(nickKey, out lastHour) && string.Equals(lastHour, hourKey, StringComparison.Ordinal))
                {
                    return;
                }

                if (_checkRunning || DateTime.UtcNow.Subtract(_lastCheckStartedAtUtc) < RetryAfterFailure)
                {
                    return;
                }

                _checkRunning = true;
                _lastCheckStartedAtUtc = DateTime.UtcNow;
            }

            ThreadPool.QueueUserWorkItem(RunCheck, new CheckState(nick, nickKey, hourKey));
        }

        private static void RunCheck(object state)
        {
            var checkState = (CheckState)state;
            bool anySuccess = false;
            int hitCount = 0;
            try
            {
                foreach (ProfessionRatingRepository.Category category in ProfessionRatingRepository.GetCategories())
                {
                    try
                    {
                        ProfessionRatingRepository.RatingTable table = ProfessionRatingRepository.LoadRating(category.Id, true);
                        anySuccess = true;
                        ProfessionRatingRepository.RatingEntry hit = FindTopTenHit(table, checkState.NormalizedNick);
                        if (hit != null)
                        {
                            hitCount++;
                            PostHitNotification(category, hit);
                        }
                    }
                    catch (Exception error)
                    {
                        ANClient.AppLog.w(TraceChain, Tag, "WEEKLY_RATING_CHECK_FAILED: id=" + category.Id.ToString(CultureInfo.InvariantCulture) + ", title=" + category.Title, error);
                    }
                }

                if (anySuccess)
                {
                    lock (SyncRoot)
                    {
                        LastCheckedHourByNick[checkState.NormalizedNick] = checkState.HourKey;
                    }
                }

                ANClient.AppLog.i(TraceChain, Tag, "WEEKLY_RATING_CHECK_DONE: nick=" + checkState.Nick + ", hour=" + checkState.HourKey + ", hits=" + hitCount.ToString(CultureInfo.InvariantCulture) + ", anySuccess=" + anySuccess);
            }
            finally
            {
                lock (SyncRoot)
                {
                    _checkRunning = false;
                }
            }
        }

        private static ProfessionRatingRepository.RatingEntry FindTopTenHit(ProfessionRatingRepository.RatingTable table, string normalizedNick)
        {
            if (table == null || normalizedNick.Length == 0)
            {
                return null;
            }

            int limit = Math.Min(10, table.Entries.Count);
            for (int index = 0; index < limit; index++)
            {
                ProfessionRatingRepository.RatingEntry entry = table.Entries[index];
                if (string.Equals(normalizedNick, ProfessionRatingRepository.NormalizeNick(entry.Nick), StringComparison.Ordinal))
                {
                    return entry;
                }
            }

            return null;
        }

        private static void PostHitNotification(ProfessionRatingRepository.Category category, ProfessionRatingRepository.RatingEntry entry)
        {
            string message = "[ProfessionRatingMonitor] Вы попали в рейтинг " + category.Title + " под № " + entry.Rank.ToString(CultureInfo.InvariantCulture) + ". Не забудьте забрать награду!";
            if (ANClient.AppVars.MainForm != null)
            {
                ANClient.AppVars.MainForm.WriteChatMsgSafe(message);
            }

            ANClient.AppLog.i(TraceChain, Tag, "WEEKLY_RATING_TOP10_NOTIFY: category=" + category.Title + ", rank=" + entry.Rank.ToString(CultureInfo.InvariantCulture) + ", nick=" + entry.Nick);
        }

        private static string ResolveCurrentNick()
        {
            return ANClient.AppVars.Profile == null || ANClient.AppVars.Profile.UserNick == null ? string.Empty : ANClient.AppVars.Profile.UserNick.Trim();
        }

        private static DateTime ResolveServerDate()
        {
            if (ANClient.AppVars.Profile != null && ANClient.AppVars.Profile.ServDiff != TimeSpan.MinValue)
            {
                return DateTime.Now.Subtract(ANClient.AppVars.Profile.ServDiff);
            }

            return ANClient.AppVars.ServerDateTime == DateTime.MinValue ? DateTime.Now : ANClient.AppVars.ServerDateTime;
        }

        private sealed class CheckState
        {
            internal CheckState(string nick, string normalizedNick, string hourKey)
            {
                Nick = nick;
                NormalizedNick = normalizedNick;
                HourKey = hourKey;
            }

            internal string Nick { get; private set; }

            internal string NormalizedNick { get; private set; }

            internal string HourKey { get; private set; }
        }
    }
}
